File: BM25_二叉树的后序遍历/java_solutions/SolutionTest.java
package BM25_二叉树的后序遍历.java_solutions;

import java.util.Arrays;

public class SolutionTest {
    public static void main(String[] args) {
        // Test Case 1: A simple binary tree with 3 nodes (Root, Left, Right)
        System.out.println("Test Case 1:");
        TreeNode root1 = new TreeNode(1);
        root1.left = new TreeNode(2);
        root1.right = new TreeNode(3);
        runTest(root1, new int[]{2, 3, 1});

        // Test Case 2: A binary tree with only one node (root only)
        System.out.println("Test Case 2:");
        TreeNode root2 = new TreeNode(1);
        runTest(root2, new int[]{1});

        // Test Case 3: A binary tree where the root has only a left child
        System.out.println("Test Case 3:");
        TreeNode root3 = new TreeNode(1);
        root3.left = new TreeNode(2);
        runTest(root3, new int[]{2, 1});

        // Test Case 4: A binary tree where the root has only a right child
        System.out.println("Test Case 4:");
        TreeNode root4 = new TreeNode(1);
        root4.right = new TreeNode(2);
        runTest(root4, new int[]{2, 1});

        // Test Case 5: A more complex tree with multiple levels (Root, Left and Right Children, and further children)
        System.out.println("Test Case 5:");
        TreeNode root5 = new TreeNode(1);
        root5.left = new TreeNode(2);
        root5.right = new TreeNode(3);
        root5.left.left = new TreeNode(4);
        root5.left.right = new TreeNode(5);
        root5.right.left = new TreeNode(6);
        root5.right.right = new TreeNode(7);
        runTest(root5, new int[]{4, 5, 2, 6, 7, 3, 1});

        // Test Case 6: An empty tree (null root)
        System.out.println("Test Case 6:");
        TreeNode root6 = null;
        runTest(root6, new int[]{});
    }
    
    // Run tests for all three solutions and compare expected with real result
    private static void runTest(TreeNode root, int[] expected) {
        StackSolution stackSolution = new StackSolution();
        int[] result = stackSolution.postorderTraversal(root);
        compareResults("StackSolution", result, expected);

        RecursiveSolution recursiveSolution = new RecursiveSolution();
        result = recursiveSolution.postorderTraversal(root);
        compareResults("RecursiveSolution", result, expected);

        BetterStackSolution betterStackSolution = new BetterStackSolution();
        result = betterStackSolution.postorderTraversal(root);
        compareResults("BetterStackSolution", result, expected);

        System.out.println();  // Newline between test cases
    }
    
    // Compare expected and real results, print "pass" or "failed"
    private static void compareResults(String solutionName, int[] result, int[] expected) {
        if (Arrays.equals(result, expected)) {
            System.out.println(solutionName + " passed");
        } else {
            System.out.println(solutionName + " failed");
            System.out.println("Expected: " + Arrays.toString(expected));
            System.out.println("Got: " + Arrays.toString(result));
        }
    }
}

File: BM25_二叉树的后序遍历/python_solutions/solution_test.py
from solution import TreeNode, Solution

# Helper function to run the test cases
def run_test_case(root, expected_output):
    solution = Solution()
    real_output = solution.postorderTraversal(root)
    
    print(f"Expected Output: {expected_output}")
    print(f"Real Output: {real_output}")
    
    if real_output == expected_output:
        print("Result: Passed\n")
    else:
        print("Result: Failed\n")


# Test Case 1: Simple Binary Tree
# Tree structure:
#     1
#    / \
#   2   3

root = TreeNode(1)
root.left = TreeNode(2)
root.right = TreeNode(3)
expected_output = [2, 3, 1]
run_test_case(root, expected_output)


# Test Case 2: Single Node Tree
# Tree structure:
#     1

root = TreeNode(1)
expected_output = [1]
run_test_case(root, expected_output)


# Test Case 3: Empty Tree
root = None
expected_output = []
run_test_case(root, expected_output)


# Test Case 4: Left-skewed Tree
# Tree structure:
#     1
#    /
#   2
#  /
# 3

root = TreeNode(1)
root.left = TreeNode(2)
root.left.left = TreeNode(3)
expected_output = [3, 2, 1]
run_test_case(root, expected_output)


# Test Case 5: Right-skewed Tree
# Tree structure:
#     1
#      \
#       2
#        \
#         3

root = TreeNode(1)
root.right = TreeNode(2)
root.right.right = TreeNode(3)
expected_output = [3, 2, 1]
run_test_case(root, expected_output)


# Test Case 6: Complex Tree
# Tree structure:
#        1
#       / \
#      2   3
#     / \    \
#    4   5    6

root = TreeNode(1)
root.left = TreeNode(2)
root.right = TreeNode(3)
root.left.left = TreeNode(4)
root.left.right = TreeNode(5)
root.right.right = TreeNode(6)
expected_output = [4, 5, 2, 6, 3, 1]
run_test_case(root, expected_output)


# Test Case 7: Tree with Only Right Subtree
# Tree structure:
#     1
#      \
#       2
#        \
#         3
#          \
#           4

root = TreeNode(1)
root.right = TreeNode(2)
root.right.right = TreeNode(3)
root.right.right.right = TreeNode(4)
expected_output = [4, 3, 2, 1]
run_test_case(root, expected_output)
